### DISCUSSION TITLE
should oreKey 1 object instead of an array?

### DISCUSSION THREAD

**Discussion by duongphuhiep** [firasdarwish/ore#11] [General]
When I first saw the example of keyed service

```go
ore.RegisterLazyCreator[Counter](ore.Scoped, &simpleCounter{}, "name here", 1234)
```

I wondered 
* is the order of "name here" and 1234 are important? may be I can `Get(1234,  "name here")`
* maybe `Get(1234)` and `Get("name here")` will give the same object...?

The current design "Key as an array" is not obvious. Without diving to the codes of ore, users might got the wrong idea. 

It would be less confuse if the key was an object `SomeKey {name: "name here", category: 1234}` instead of an array `["name here", 1234]`.

However if we remove the variadic `...` then we will have to add many func `RegisterKeyedXXX` because there is no func overloading in Go.

IMO, removing variadic keys willl make the API more explicit / obvious, We don't even need to serialize key to string as long as the key is constraint to an "Equality comparable" type.

The tradeoff is a new bunch of func `RegisterKeyedXXX` will appear in Ore, However it is just the "Go-way" having a bunch of functions with slightly different name.

**Comment by firasdarwish** [comment]
you're correct, this was bugging a long time ago, unfortunately it is because of Go's lack of overloading nor method-arg default values

i thought of removing the key arg altogether and make all registration funcs return an interface which allows the developer to optionally set the key

```go
ore.RegisterLazyFuncToContainer(container, ore.Transient, func(ctx context.Context) (*A, context.Context) {
		b, ctx := ore.GetFromContainer[*B](container, ctx)
		c, ctx := ore.GetFromContainer[*C](container, ctx)
		return NewA(b, c), ctx
	}).WithKey("some-key-here")
```

but this would require much re-work of how registration is done

i guess your suggestion of make new funcs for registration with `Keyed` in the names would be better. this also happens in Dotnet

```cs
builder.Services.AddKeyedSingleton<INotificationService, SmsNotificationService>("sms");
```

**Reply by duongphuhiep** [reply]
yes, the builder pattern ".WithKey" will requires more work to the registration, along with some complexity for users, what if user do a bunch of RegisterXXX first then do a bunch of WithKey() later (in other place), his co-worker won't exactly like it.

**Comment by duongphuhiep** [comment]
Now with the 0.6.0 out. 

1) I'd suggest to use directly the original key without converting it to string:

```
type typeID struct {
	pointerTypeName pointerTypeName
	oreKey          KeyStringer //replace "string" with KeyStringer here
}
```

The Pros is that we can remove all the string serialization codes in `key.go` (so better perf as a result)
The Cons is that the KeyStringer type would be limited to "comparable" only, we  would no longer be able to use slice as Key

(I didn't try yet, it is all just my estimation)

-----------

2) Speaking about `comparable` 

Is there any reason for T not to be "any" in this case?

```go
RegisterSingleton[T comparable]
ProvideScopedValue[T comparable]
```

**Reply by firasdarwish** [reply]
changes are now live in v0.6.1
